### ISSUE TITLE
Provide current webpacker instance as an option

### ISSUE THREAD

**Issue by ferblape** [rails/webpacker#2946]
I'm working on migrating a huge engine from our beloved asset pipeline to webpacker. It's going quite smoothly (thanks for your work and your engines guide) except a little block I've found: the engine defines the layout of the application, and our plan is let the application to declare a pack "application" while the engine will pack all its javascript in a pack called "engine_name", so both packs are declared in the layout together:

```
<%= javascript_pack_tag "engine_name" %>
<%= javascript_pack_tag "application" %>
```

While I define in my helpers `current_webpacker_instance` ([step#5](https://github.com/rails/webpacker/blob/5-x-stable/docs/engines.md#step-5-configure-helper)) it finds the engine_name pack but not the application pack.

I'm considering adding the current instance pack as an option to the helpers. Any opinions about this?

**Comment by krists** [comment]
This requested feature would be very useful in scenarios where Rails engine and main application share the same layout and needs to include assets from both sources.

Here is an example:

* Main application layout file that renders basic document structure and navigation partial.
  ```erb
  <% # Main app's app/views/layouts/application.html.erb %>
  <html>
    <head>
      <%= stylesheet_link_tag 'application' %>
      <%= javascript_pack_tag 'application' %>
      <%= yield :head_assets %>
    </head>
    <body>
      <%= render "navigation" %>
      <%= content_for?(:secondary_layout) ? yield(:secondary_layout) : yield %>
    </body>
  </html>
  ```
* Engine layout that inherits from main application layout.
  ```erb
  <% # Engine app's app/views/layouts/my_engine/application.html.erb %>
  <% content_for :head_assets do %>
    <%= stylesheet_link_tag "my_engine/application" %>
    <%= javascript_pack_tag 'my_engine/application' %>
  <% end %>
  <% content_for :secondary_layout do %>
    <%= yield %>
  <% end %>
  <%= render template: "layouts/application" %>
  ```
Unfortunately this does not work because `current_webpacker_instance`  lookup method will fail to find "my_engine/application.js" file.

One way how to solve it is to temporary change `Webpacker.webpacker` class variable with engine's instance.
```erb
module MyEngine
  ROOT_PATH = Pathname.new(File.expand_path(File.join(__dir__, "..")))
  class << self
    def webpacker
      @webpacker ||= ::Webpacker::Instance.new(
        root_path: ROOT_PATH,
        config_path: ROOT_PATH.join("config/webpacker.yml")
      )
    end

    def with_my_engine_webpacker_instance
      original_instance = Webpacker.instance
      Webpacker.instance = self.webpacker
      yield(self.webpacker, original_instance)
    ensure
      Webpacker.instance = original_instance
    end
  end
end

```

And then use it in template like this:
```erb
<% MyEngine.with_my_engine_webpacker_instance do %>
  <%= stylesheet_link_tag "my_engine/application" %>
  <%= javascript_pack_tag 'my_engine/application' %>
<% end %>
```

But this a bad example as class variables are not thread-safe.

It would be  much better to have a instance parameter for all [helpers](https://github.com/rails/webpacker/blob/5-x-stable/lib/webpacker/helper.rb) like `javascript_pack_tag` or `javascript_packs_with_chunks_tag`.

MyEngine layout then would look something like this:
```erb
<% content_for :head_assets do %>
  <%= stylesheet_link_tag "my_engine/application", instance: MyEngine.webpacker %>
  <%= javascript_pack_tag 'my_engine/application', instance: MyEngine.webpacker %>
<% end %>
<% content_for :secondary_layout do %>
  <%= yield %>
<% end %>
<%= render template: "layouts/application" %>
```

What are your thoughts @gauravtiwari ? If you agree on this change then  I would be happy to help and open a pull-request.

**Comment by ferblape** [comment]
That's exactly my scenario, I'm working in a very big engine where I mix the webpacker instance of the engine with the assets of the container app, which can be in webpacker or not.

**Comment by kyrofa** [comment]
I'm hitting this as well. My use-case is very typical: a JS-heavy view from an engine being rendered within a layout provided by the main app. The main app also has some JS for the layout, while the JS in the engine is specific to the functionality provided by the engine. I was trying to figure out why my navigation was broken when I finally realized that my `javascript_pack_tag 'application'` in the layout was picking up the pack from the engine.

I agree that the proposed solution would fix this situation, but I would add that the default should depend on the view context: default to the webpacker from the engine that is rendering the view (treating the main app as just another engine, because of course, it is).

**Comment by kyrofa** [comment]
Hey @palkan, any chance you know the current status of engine support in webpacker? The engine doc has disappeared in master, and following the 5-x engine doc while using master doesn't work today. I was going to try to write and propose a solution to this, but I want to propose that fix to master and potentially backport as opposed to proposing to 5-x where engines work (barring this issue). I just need to get master actually _working_, first.

**Comment by kyrofa** [comment]
Went ahead and took a crack at a basic fix for this in #3072.
